session/headerjar.go

```diff
@@ -1,30 +1,67 @@
 package session
 
 import (
+	"fmt"
 	"net/http"
 	"sync"
 )
 
 // HeaderJar map between hosts and headers
 type HeaderJar struct {
-	headers *sync.Map
+	headers        *sync.Map
+	allRestMethods []RestMethod
+}
+
+// hostMethodPair acts as a key in the map
+type hostMethodPair struct {
+	host   string
+	method RestMethod
+}
+
+// String returns string interpretation of this hostMethodPair
+func (hmp *hostMethodPair) String() string {
+	return fmt.Sprintf("%s:%s", hmp.host, hmp.method.String())
 }
 
 // NewHeaderJar returns an empty HeaderJar
 func NewHeaderJar() *HeaderJar {
+	restMethodInts := RestMethod(0).GetEnumMap().AsInt()
+	allRestMethods := make([]RestMethod, 0, len(restMethodInts))
+	for _, restMethod := range restMethodInts {
+		allRestMethods = append(allRestMethods, RestMethod(restMethod))
+	}
+
 	return &HeaderJar{
-		headers: &sync.Map{},
+		headers:        &sync.Map{},
+		allRestMethods: allRestMethods,
 	}
 }
 
 // SetHeader set header for a particular host
 func (hj *HeaderJar) SetHeader(host string, header http.Header) {
-	hj.headers.Store(host, header)
+	if len(hj.allRestMethods) == 0 {
+		panic("HeaderJar may only be initialized using NewHeaderJar()")
```

Comment:
never like seeing `panic` as it will affect all users, but guess the purpose if this is to catch programming errors early.